crates/bevy_ecs/src/system/exclusive_system.rs

```diff
@@ -15,59 +18,77 @@ pub trait ExclusiveSystem: Send + Sync + 'static {
-where
-    F: FnMut(&mut World) + Send + Sync + 'static,
-{
-    fn name(&self) -> Cow<'static, str> {
-        self.name.clone()
-    }
+macro_rules! impl_exclusive_system {
+    ($($t:ident),*) => {
+        #[allow(non_snake_case)]
+        impl<F, $($t: FromWorld + Resource,)*> ExclusiveSystem for ExclusiveSystemFn<F, ($($t,)*)>
+        where
+            F: FnMut(&mut World, $(Local<$t>,)*) + Send + Sync + 'static,
+        {
+            fn name(&self) -> Cow<'static, str> {
+                self.name.clone()
+            }
 
-    fn run(&mut self, world: &mut World) {
-        // The previous value is saved in case this exclusive system is run by another exclusive
-        // system
-        let saved_last_tick = world.last_change_tick;
-        world.last_change_tick = self.last_change_tick;
+            fn run(&mut self, world: &mut World) {
+                // The previous value is saved in case this exclusive system is run by another exclusive
+                // system
+                let saved_last_tick = world.last_change_tick;
+                world.last_change_tick = self.last_change_tick;
 
-        (self.func)(world);
+                let ($($t,)*) = self.locals.as_mut().unwrap();
+                (self.func)(world, $(Local::wrap($t),)*);
 
-        let change_tick = world.change_tick.get_mut();
-        self.last_change_tick = *change_tick;
-        *change_tick += 1;
+                let change_tick = world.change_tick.get_mut();
+                self.last_change_tick = *change_tick;
+                *change_tick += 1;
 
-        world.last_change_tick = saved_last_tick;
-    }
+                world.last_change_tick = saved_last_tick;
+            }
 
-    fn initialize(&mut self, _: &mut World) {}
+            fn initialize(&mut self, _world: &mut World) {
+                self.locals = Some(($($t::from_world(_world),)*));
+            }
 
-    fn check_change_tick(&mut self, change_tick: u32) {
-        check_system_change_tick(&mut self.last_change_tick, change_tick, self.name.as_ref());
-    }
+            fn check_change_tick(&mut self, change_tick: u32) {
+                check_system_change_tick(
+                    &mut self.last_change_tick,
+                    change_tick,
+                    self.name.as_ref(),
+                );
+            }
+        }
+
+
+        impl<F, $($t: FromWorld + Resource,)*> IntoExclusiveSystem<&mut World, ExclusiveSystemFn<F, ($($t,)*)>> for F
+        where
+            F: FnMut(&mut World, $(Local<$t>,)*) + Send + Sync + 'static,
+        {
+            fn exclusive_system(self) -> ExclusiveSystemFn<F, ($($t,)*)> {
+                ExclusiveSystemFn {
+                    func: self,
+                    name: core::any::type_name::<F>().into(),
+                    last_change_tick: 0,
+                    locals: None,
+                }
+            }
+        }
+
+    };
 }
 
+all_tuples!(impl_exclusive_system, 0, 12, L);
```

Comment:
```suggestion
all_tuples!(impl_exclusive_system, 0, 16, L);
```
I would rather have all all_tuples work on the same maximum number of items in a tuple, so 16. It would be less surprising for users